ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffsetFactory.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udaf.offset;
+
+import com.google.common.collect.ImmutableList;
+import io.confluent.ksql.function.AggregateFunctionFactory;
+import io.confluent.ksql.function.AggregateFunctionInitArguments;
+import io.confluent.ksql.function.KsqlAggregateFunction;
+import io.confluent.ksql.function.UdafAggregateFunctionFactory;
+import io.confluent.ksql.function.UdfIndex;
+import io.confluent.ksql.function.types.ParamType;
+import io.confluent.ksql.schema.ksql.SchemaConverters;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlArray;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.KsqlException;
+import java.util.List;
+import org.apache.kafka.connect.data.Schema;
+
+public class EarliestByOffsetFactory extends AggregateFunctionFactory {
+  static final String NAME = "EARLIEST_BY_OFFSET";
+
+  public EarliestByOffsetFactory() {
+    super(NAME);
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  @Override
+  public KsqlAggregateFunction createAggregateFunction(
+      final List<SqlArgument> argTypeList,
+      final AggregateFunctionInitArguments initArgs
```

Comment:
What's the difference between these two arguments? I'm getting confused about them.

Reply:
Yeah, they are confusing....

We need an example like `earliest_by_offset(column, n)` to see the difference.  

Here `column` would be passed the `argTypeList`.  In the case of UDAFs there is a current restriction/assumption that UDAFS may only act on one column.  So `argTypeList` will usually be a singleton.  (It won't have more elements.  I cannot remember how `CountKudaf` works precisely.  Maybe that is empty...)

On the other hand, `initArgs` contains everything else one would need to know in order to make the UDAF go.  In this specific example, the value `n` would be in an array of arguments.  Additionally there's configuration Map which can hold whatever one likes which is passed to Udafs which implement `Configurable`.  

As an example of a `Configurable` UDAF, `collect_list` and `collect_set` have limits which are configured at the system level to limit the size of the collection maintained in memory.

One option I considered is adding the `argTypeList` in the configuration map.  This would work, but it seems a little brittle....